### ISSUE TITLE
trying to train a text classifier with ELMO embeddings but accuracy on test doesn't improve

### ISSUE THREAD

**Issue by GiliGoldin** [flairNLP/flair#709]
I have a dataset which contains chunks of 100 **tokenized** sentences written in English by users on Reddit. Each chunk contains only sentences written by the same user. The labels in the dataset are the native language of the users (30 different labels). 
I wanted to train a text classifier that predicts the native language for each chunk.
I wanted to do this using ELMO embeddings (both with a default ELMO model and with a self-trained one)
I tried to follow the tutorials.
I processed the dataset to be in the following format: 

> __label__<label_1> < text>

 as was written in [tutorial_6](https://github.com/zalandoresearch/flair/blob/master/resources/docs/TUTORIAL_6_CORPUS.md)
where <label_1> will be for instance Spanish and < text> will be one full chunk.
The problem is that the model is not improving on the dev and test set, although I do see the loss going down from epoch to epoch on the training data.

I tried a lot of variations on the data processing such as: 
- smaller chunks (50 sentences instead of 100)
- removing punctuations
- adding < S>,< /S> signs between sentences
- using only 5000 most frequent tokens and replacing the rest with 'OOV' etc.

But nothing worked.
Even when I made the task simpler by using only 5 'language families' as labels instead of 30 different languages, or even simpler by using binary labels for 'native English speakers' and 'non-native English speakers', the results were more or less the same. 

I'll just mention that I was able to create a successful model (A simple LSTM with bahdanau attention and self-trained embeddings) using Keras, on the same data.

Do you happen to have any suggestions on what I am doing wrong?
 
This is my code:

```python
import os

from flair.data import TaggedCorpus, Sentence
from flair.data_fetcher import NLPTaskDataFetcher, NLPTask
from flair.embeddings import WordEmbeddings, FlairEmbeddings, DocumentRNNEmbeddings, ELMoEmbeddings
from flair.models import TextClassifier
from flair.trainers import ModelTrainer
from flair.visual.training_curves import Plotter
from flair.training_utils import EvaluationMetric


from prepare_corpus import ONLY_EUROPE_CORPUS_PATH, NON_EUROPE_CORPUS_PATH 
ONLY_EUROPE = True
ELMO_default = True
ELMO_reddit_native = False
Elmo_reddit_non_native = False


if ONLY_EUROPE:
    corpus_path = ONLY_EUROPE_CORPUS_PATH
    base_path = 'resources/taggers/only_europe_reddit'
else:
    corpus_path = NON_EUROPE_CORPUS_PATH
    base_path = 'resources/taggers/non_europe_reddit'

import torch
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using device:', device)
print()

#Additional Info when using cuda
 if device.type == 'cuda':
    print(torch.cuda.get_device_name(0))
    print('Memory Usage:')
    print('Allocated:', round(torch.cuda.memory_allocated(0)/1024**3,1), 'GB')
    print('Cached:   ', round(torch.cuda.memory_cached(0)/1024**3,1), 'GB')

# 1. get the corpus
 corpus: TaggedCorpus = NLPTaskDataFetcher.load_classification_corpus(corpus_path, test_file='test.txt', train_file='train.txt', use_tokenizer=False)

# 2. create the label dictionary
 label_dict = corpus.make_label_dictionary()

# 3. make a list of word embeddings
 if ELMO_default:
    word_embeddings = [ELMoEmbeddings()]
    base_path = base_path + '_default_ELMO'
elif ELMO_reddit_native:
    word_embeddings = [ELMoEmbeddings('custom', options_file = "elmo_files/native_options.json", weight_file= "elmo_files/native_weights.hdf5")]
    base_path = base_path + '_reddit_native_ELMO'
elif Elmo_reddit_non_native:
    word_embeddings = [ELMoEmbeddings('custom', options_file = "elmo_files/nonnative_options.json", weight_file= "elmo_files/nonnative_weights.hdf5")]
    base_path = base_path + '_reddit_non_native_ELMO'

# 4. initialize document embedding by passing list of word embeddings
document_embeddings: DocumentRNNEmbeddings = DocumentRNNEmbeddings(word_embeddings,  
                                                                     hidden_size=128,
                                                                     # reproject_words=True,           
                                                                     # reproject_words_dimension=256,
                                                                     rnn_type='LSTM',
                                                                     dropout=0.5
                                                                     ) #(I tried different variations of these params)

# 5. create the text classifier
classifier = TextClassifier(document_embeddings, label_dictionary=label_dict, multi_label=False)


# 6. initialize the text classifier trainer
trainer = ModelTrainer(classifier, corpus)
# 7. start the training
trainer.train(base_path,
                  learning_rate=0.1,
                  mini_batch_size=20,
                  anneal_factor=0.5,
                  patience=3,
                  max_epochs=100, evaluation_metric=EvaluationMetric.MACRO_ACCURACY, embeddings_in_memory=True)
```

Thank you!

**Comment by alanakbik** [comment]
Hello @GiliGoldin that sounds interesting. 

Quick question: Are you using a multilingual ELMo model for this? Most ELMo models I'm aware of are for English only, so using them on multilingual text will probably not work. 

You could try the multilingual Flair models instead - they were trained over 6 languages but were shown to work even for other European languages they weren't trained on (see [paper](http://alanakbik.github.io/papers/nldl2019.pdf)). Initialize like this: 

```python
word_embeddings = [FlairEmbeddings('multi-forward'), FlairEmbeddings('multi-backward') ]
```
Could you try and see if this works better?

**Comment by GiliGoldin** [comment]
Hi @alanakbik thank's for the reply!
Actually, all the texts in the dataset are only in English, so i didn't use multilingual models.
The chunks contain English sentences that were written by non-native English speakers (and some also by native English speakers). 
I try to predict the native tongue of the writers, by using only their posts in English.
For example, this is a partial row from the dataset:

> __label__France   because you can live your entire life with no major issue while only knowing one language .    people probably consider it 's not worth the effort .    i love languages but i understand where they come from . (...)

Hope this is more clear now

**Comment by alanakbik** [comment]
Thanks for the clarification! I would probably not do any preprocessing and just use the text as-is since this will likely lead to better ELMo embeddings. 

But it is strange that the model doesn't learn - your code looks good. Can you share the training log files? Is the dataset available somewhere?

**Comment by GiliGoldin** [comment]
Thanks for taking the time and looking into this!
I uploaded the dataset to google drive, you can find the unprocessed version here: https://drive.google.com/open?id=1n9UGegXYB8QVQeX7rjaOyQ03ZywXaZWX

I only have the log for the 'language family' version (attached), where I changed the labels from countries/languages to language-families (5 labels instead of 30 different labels)
In this case, there was an improvement on the test set in the beginning but then it stopped. in the other cases the results on the dev and test set didn't improve at all from epoch to epoch. Most of the times it stayed exactly the same number and it was much lower than chance.
[training.log](https://github.com/zalandoresearch/flair/files/3157988/training.log)

**Comment by alanakbik** [comment]
@DavidMortensen Yes that is strange - as a sanity check, what happens if you use only normal `WordEmbeddings('en')` and not ELMo in this case?

**Comment by GiliGoldin** [comment]
@alanakbik I also tried to use the normal word embedding and got the same results. I'm attaching the log.
[training.log](https://github.com/zalandoresearch/flair/files/3191632/training.log)

**Comment by alanakbik** [comment]
@GiliGoldin thanks for sharing this. Could you try reducing the mini-batch size from 20 to something really low, like 2 or 4?

**Comment by giligold** [comment]
@alanakbik Ok, I reduced the mini-batch to 2, I'm attaching the log file:
[training.log](https://github.com/zalandoresearch/flair/files/3205810/training.log)

I also tried the character embeddings (with mini batch of 12) :
[training_char_embedding.log](https://github.com/zalandoresearch/flair/files/3205817/training_char_embedding.log)

**Comment by alanakbik** [comment]
Thanks for sharing this - I'm still not sure why the model is not learning so well. One theory is that it may have something to do with learning word embeddings for this task (you mentioned your Keras model has learnable embeddings). In our implementation, word embeddings are frozen and only the document embeddings are learnt. Maybe this fails in a task like this one where the target semantics (i.e. which words are used by native speakers of different languages) are so different from pre-trained word embeddings that it is better to learn new embeddings or at least fine-tune existing embeddings. 

For this reason, I've added learnable embeddings in a [branch](https://github.com/zalandoresearch/flair/tree/GH-678-classification) that I will push to master soon(ish). Perhaps we could try these out on your data?

Could you also share the code of your Keras model so we can have a better look at a model that works on this data?

**Comment by GiliGoldin** [comment]
@alanakbik 
Sure, here is the code of my Keras model: https://github.com/GiliGoldin/keras_reddit_NLI
I couldn't find the learnable embeddings in the branch you provided, could you direct me to the relevant functions? thank's!

**Comment by alanakbik** [comment]
Thanks! The new class is the `OneHotEmbeddings` class, a good baseline to try would be: 

```python
# instantiate one-hot encoded word embeddings with your corpus
embeddings_1 = OneHotEmbeddings(corpus)

# instantiate word embeddings that work well for you
embeddings_2 = WordEmbeddings('glove')

# document pool embeddings
document_embeddings = DocumentPoolEmbeddings([embeddings_1, embeddings_2], fine_tune_mode='none')
```

Should this work better than the other approaches this would support the theory that we need learnable embeddings to do this type of classification task.
